src/services/event-service.js

```diff
@@ -0,0 +1,202 @@
+import NotificationStore from '../stores/notification-store.js';
+import TransactionRunner from '../stores/transaction-runner.js';
+import DocumentRevisionStore from '../stores/document-revision-store.js';
+import {
+  determineNotificationReasonsForCommentCreatedEvent,
+  determineNotificationReasonsForRevisionCreatedEvent
+} from '../utils/notification-utils.js';
+
+const logger = new Logger(import.meta.url);
+
+const MAX_PROCESSING_ATTEMPTS = 3;
+
+const PROCESSING_RESULT = {
+  succeeded: 'succeeded',
+  cancelled: 'cancelled',
+  failed: 'failed'
+};
+
+class EventService {
+  static dependencies = [EventStore, NotificationStore, UserStore, DocumentStore, DocumentRevisionStore, RoomStore, LockStore, TransactionRunner];
+
+  constructor(eventStore, notificationStore, userStore, documentStore, documentRevisionStore, roomStore, lockStore, transactionRunner) {
+    this.eventStore = eventStore;
+    this.notificationStore = notificationStore;
+    this.userStore = userStore;
+    this.documentStore = documentStore;
+    this.documentRevisionStore = documentRevisionStore;
+    this.roomStore = roomStore;
+    this.lockStore = lockStore;
+    this.transactionRunner = transactionRunner;
+  }
+
+  async _processRevisionCreatedEvent(event, session, context) {
+    const userIterator = this.userStore.iterateAllActiveUsers({ session });
+
+    try {
+      const { revisionId, documentId, roomId } = event.params;
+      const revision = await this.documentRevisionStore.getDocumentRevisionById(revisionId, { session });
+      const document = await this.documentStore.getDocumentById(documentId, { session });
+      const room = roomId ? await this.roomStore.getRoomById(roomId, { session }) : null;
+
+      const createdNotifications = [];
+      for await (const user of userIterator) {
+        if (context.cancellationRequested) {
+          return PROCESSING_RESULT.cancelled;
+        }
+
+        const reasons = determineNotificationReasonsForRevisionCreatedEvent({
+          event,
+          revision,
+          document,
+          room,
+          notifiedUser: user
+        });
+
+        if (reasons.length) {
+          createdNotifications.push({
+            _id: uniqueId.create(),
+            notifiedUserId: user._id,
+            eventId: event._id,
+            eventType: event.type,
+            eventParams: event.params,
+            reasons,
+            createdOn: event.createdOn,
+            readOn: null
+          });
+        }
+      }
+
+      if (createdNotifications.length) {
+        await this.notificationStore.insertNotifications(createdNotifications, { session });
+      }
+
+      return PROCESSING_RESULT.succeeded;
+    } finally {
+      await userIterator.close();
+    }
+  }
+
+  async _processCommentCreatedEvent(event, session, context) {
+    const userIterator = this.userStore.iterateAllActiveUsers({ session });
+
+    try {
+      const { documentId, roomId } = event.params;
```

Comment:
Actually we don't even need the `roomId` in this case, because we have it on the `document`. It's even maybe room for discrepancy if we pass it separately? 🤔

Reply:
Hm, that's true :thinking:

On the other hand, then we could also argue that we do not need the `documentId` as it is already on the revision (at least for the `revision-created` type of event). It would make things harder though when assembling the all the data needed to show user notifications (as the notifications use the same event params). We would have to do multiple round of fetches to the db. So I can see both a case for removing and keeping the redundant data. Tending more towards keeping. What do you think?

Reply:
Ok then, let's keep them.